### PR TITLE
Admin UI Update

### PR DESCRIPTION
**Pull request by kldavis4**
This PR moves admin / reports to the react UI. This reduces complexity in this project and it is a step towards adding file copy functionality in the admin UI.
This also fixes a bug in the python admin api where it wasn't calling the admin script for the system calls (reboot, shutdown, etc).

### CODE REVIEW

File: ansible/roles/nginx/templates/connectbox_static-site.conf.j2 [hunk 1]
```diff
@@ -11,4 +11,9 @@ server {
 {{ nginx_chat_block }}
 {{ nginx_admin_block }}
 
+    # Admin interface
+    location /admin {
```
Comment (edwinsteele) [inline ansible/roles/nginx/templates/connectbox_static-site.conf.j2]: The `auth_basic` and `auth_basic_user_file` parameters are missing here. I notice that the tests passed, and knowing one of them tests for a 404, I suspect somethign else is doing the auth. Is that correct? Can you please add a comment to that end if it's the case.

File: ansible/roles/webserver-content/tasks/main.yml [hunk 2]
```diff
@@ -20,33 +20,13 @@
   git:
     repo: "{{ connectbox_client_repo }}"
     dest: /tmp/connectbox_client
+    version: react_admin_ui
```
Comment (edwinsteele) [inline ansible/roles/webserver-content/tasks/main.yml]: This will need to be removed before merging.

File: python/admin/api.py [hunk 3]
```diff
@@ -50,10 +50,8 @@ def set_system_property():
     if (not request.json) or ("value" not in request.json):
         _abort_bad_request() # bad request
 
-    if request.json["value"] in ["shutdown", "reboot"]:
-        # Safe as input is limited
-        subprocess.Popen("sleep 1; " + request.json["value"], shell=True)
-        return jsonify(code=0, result="SUCCESS")
+    if not request.json["value"] in ["shutdown", "reboot", "unmountusb", "reset"]:
+        _abort_bad_request() # bad request
```
Comment (edwinsteele) [inline python/admin/api.py]: The `sleep 1` was here, for memory, because it allowed us to get a success/fail result back from the API. Why the change?

File: tests/test_connectbox_static.py [hunk 4]
```diff
@@ -70,35 +70,6 @@ def getAdminAuth():
     return requests.auth.HTTPBasicAuth(ADMIN_USER, ADMIN_PASSWORD)
 
 
-class ConnectBoxBasicTestCase(unittest.TestCase):
-
-    def testContentResponseType(self):
```
Comment (edwinsteele) [inline tests/test_connectbox_static.py]: why is this being removed? It's exercising the webserver rather than the interface code, so I'd expect it to stay.

File: tests/test_connectbox_static.py [hunk 5]
```diff
@@ -70,35 +70,6 @@ def getAdminAuth():
...
-        # URLs under content should return json
-        r = requests.get("%s/content/" % (getTestBaseURL(),))
-        r.raise_for_status()
-        self.assertIsInstance(r.json(), list)
-
-    def testAdminNeedsAuth(self):
```
Comment (edwinsteele) [inline tests/test_connectbox_static.py]: why is this being removed? It's exercising the webserver rather than the interface code, so I'd expect it to stay.

File: tests/test_connectbox_static.py [hunk 6]
```diff
@@ -70,35 +70,6 @@ def getAdminAuth():
...
-        r = requests.get("%s/" % (getAdminBaseURL(),))
-        # No raise_for_status because we're checking for a 401
-        self.assertEqual(r.status_code, 401)
-
-    def testAdminNoTrailingSlashRequired(self):
```
Comment (edwinsteele) [inline tests/test_connectbox_static.py]: More general comment, given I've just added two specific comments for functions above. Why are any of these tests being removed? :-)

File: ansible/roles/nginx/templates/connectbox_static-site.conf.j2 (hunk 1 above)
Comment (kldavis4) [reply ansible/roles/nginx/templates/connectbox_static-site.conf.j2]: `/admin` is no longer protected since it is hosted by the unified UI. The only protected url is the `/admin/api.py` path: https://github.com/ConnectBox/connectbox-pi/blob/master/ansible/group_vars/all#L60

File: ansible/roles/webserver-content/tasks/main.yml (hunk 2 above)
Comment (kldavis4) [reply ansible/roles/webserver-content/tasks/main.yml]: That's correct. This is just there so that this will pull the branch version of the UI until I release the it to master.

File: python/admin/api.py (hunk 3 above)
Comment (kldavis4) [reply python/admin/api.py]: So apparently what that code was doing was calling shutdown/reboot directly instead of passing the command through to the ConnectBoxManage.sh script. That differs from the php admin script and I don't think that's what we want. At the moment these functions don't work correctly because of that.

File: tests/test_connectbox_static.py (hunk 4 above)
Comment (kldavis4) [reply tests/test_connectbox_static.py]: I removed these because /admin/ is no longer doing basic auth and the /admin/ path is handled by the react app now instead of the web server. I should probably add a test for /admin/api.py to validate basic auth, but the rest of these no longer apply.

**Comment by edwinsteele** [comment]
Unrelated to previous comments, the change password function doesn't work on firefox i.e. the password isn't changed (confirmed by looking at the contents of `/usr/local/connectbox/etc/basicauth` and the lack of re-prompting for password).

From firefox:
capture1

From Chrome:
chrome-capture

**Comment by kldavis4** [comment]
Very interesting.. I made some updates this morning which seemed to be the culprit. Just not sure why, given it seems to be working in chrome

**Comment by edwinsteele** [comment]
Looking at the request headers, firefox has a bunch of cache control headers (and the other doesn't). Perhaps that's related?

Re: tests. I think the following still provide value:
* testContentResponseType: confirms the webserver is serving json under /content/ (catch webserver misconfig)
* testAdminNeedsAuth: with a new path, confirming that we're indeed protecting the admin api
* testAdminNoTrailingSlashRequired: confirms that /admin redirects to /admin/ for people typing it in (this is currently broken, in that /admin hangs, displaying "Authenticating...")
* testAdminPageTitle: smoketest that the admin URL used in the settings link works.

File: python/admin/api.py (hunk 3 above)
Comment (edwinsteele) [reply python/admin/api.py]: great. thanks.

**Comment by kldavis4** [comment]
Regarding the password issue in firefox, I think this is a bit more complicated than I initially thought. It appears that even in Chrome, the PUT request to update the password gets cancelled. The request seems to still make it through and update the password file, so it still 'works' in Chrome. In firefox it looks like the request doesn't make it through consistently, so the password file is not updating. 

If I take out everything but the update to /usr/local/connectbox/etc/basicauth I don't get the cancelled request, so it looks like nginx detects the file update and aborts the request. Not sure what the best solution for this is. Here are some ideas and would love your input:

a) Have some kind of 'new password' file that gets created by set_password and have that get moved into place by the nginx init script if that file exists.
b) Implement some kind of custom auth (not basic auth) in the application layer. Would require some backend work in the python api but would not require any reload of nginx or gunicorn to handle password updates.

**Comment by kldavis4** [comment]
well it turns out, after spending a couple days troubleshooting, the password problem was on the client :( I used an html form without thinking too much about it and it was submitting the form instead of finishing the ajax request. Real tough to figure out because it looked like an upstream issue (gunicorn). So anyway, sort of good news, and hopefully I can make progress on the other issues...

**Comment by edwinsteele** [comment]
Looks like the build is passing. I think everything's been addressed, so unless you know of other critical issues, I think we should get this in.
